src/PowerShellEditorServices/Language/RunspaceSynchronizer.cs

```diff
@@ -0,0 +1,272 @@
+
+        private readonly static HashSet<string> POWERSHELL_MAGIC_VARIABLES = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "PID",
+            "PSVersionTable",
+            "PSEdition",
+            "PSHOME",
+            "HOST",
+            "true",
+            "false",
+            "null",
+            "Error",
+            "IsMacOS",
+            "IsLinux",
+            "IsWindows"
+        };
+
+        /// <summary>
+        /// Determines if the RunspaceSynchronizer has been initialized.
+        /// </summary>
+        public static bool IsReadyForEvents { get; private set; }
+
+        #region Public methods
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void InitializeRunspaces(Runspace runspaceSource, Runspace runspaceTarget)
+        {
+            sourceRunspace = runspaceSource;
+            sourceEngineIntrinsics = ReflectionUtils.GetEngineIntrinsics(sourceRunspace);
+            targetRunspace = runspaceTarget;
+            targetEngineIntrinsics = ReflectionUtils.GetEngineIntrinsics(runspaceTarget);
+            IsReadyForEvents = true;
+
+            sourceEngineIntrinsics.Events.SubscribeEvent(
+                source: null,
+                eventName: null,
+                sourceIdentifier: PSEngineEvent.OnIdle.ToString(),
+                data: null,
+                handlerDelegate: HandleRunspaceStateChange,
+                supportEvent: true,
+                forwardEvent: false);
+
+            Activate();
+            // Trigger events
+            HandleRunspaceStateChange(sender: null, args: null);
+        }
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void Activate()
+        {
+            SourceActionEnabled = true;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static void HandleRunspaceStateChange(object sender, PSEventArgs args)
+        {
+            if (!SourceActionEnabled)
+            {
+                return;
+            }
+
+            SourceActionEnabled = false;
+
+            var newOrChangedModules = new List<PSModuleInfo>();
+            List<PSModuleInfo> modules = ReflectionUtils.GetModules(sourceRunspace);
+            foreach (PSModuleInfo module in modules)
+            {
+                if (moduleCache.Add(module))
+                {
+                    newOrChangedModules.Add(module);
+                }
+            }
+
+
+            var newOrChangedVars = new List<PSVariable>();
+
+            var variables = sourceEngineIntrinsics.GetVariables();
+            foreach (var variable in variables)
+            {
+                // If the variable is a magic variable or it's type has not changed, then skip it.
+                if(POWERSHELL_MAGIC_VARIABLES.Contains(variable.Name) ||
+                    (variableCache.TryGetValue(variable.Name, out Type value) && value == variable.Value?.GetType()))
```

Comment:
Completion isn't just done by type, member inference of a property expression is done by value (which would also be an issue for properties with a thread static backing field). This also doesn't account for `PSObject`'s that may have gained additional members.

Reply:
> Completion isn't just done by type, member inference of a property expression is done by value ... This also doesn't account for PSObject's that may have gained additional members.

True, absolutely valid concerns. Maybe check on the object equality instead of type equality here.

Reply:
> Maybe check on the object equality instead of type equality here.

You could do that, but thread static is still a concern. 

Another issue to consider is any properties that attempt to marshal the call back to their origin thread, such as `CommandInfo.Parameters`.

Reply:
now checking object equality